### PR TITLE
[improvement](cloud) Accelerate creating table by batching RPC In cloud

### PR DESCRIPTION
**Pull request by deardeng**
Create dynamic partitions, optimize from calling prepare rpc and commit
rpc multiple times to only calling prepare and commit once

```
in one server(fe/ms)

before optimize：
Query OK, 0 rows affected (16.96 sec)
Query OK, 0 rows affected (16.65 sec)
Query OK, 0 rows affected (16.26 sec)
Query OK, 0 rows affected (16.39 sec)
Query OK, 0 rows affected (17.15 sec)

after optimize：
Query OK, 0 rows affected (13.31 sec)
Query OK, 0 rows affected (11.71 sec)
Query OK, 0 rows affected (11.81 sec)
Query OK, 0 rows affected (11.55 sec)
Query OK, 0 rows affected (12.07 sec)

......
734 rows in set (1.06 sec)
```

### CODE REVIEW

**Comment by deardeng** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

File: cloud/src/meta-service/meta_service.h [hunk 1]
```diff
@@ -280,6 +283,13 @@ class MetaServiceImpl : public cloud::MetaService {
             const AlterInstanceRequest* request,
             std::function<std::pair<MetaServiceCode, std::string>(InstanceInfoPB*)> action);
 
+    using check_create_table_type = std::function<const std::tuple<
+            const ::google::protobuf::RepeatedField<int64_t>, std::string,
+            std::function<std::string(std::string, int64_t)>>(const CheckKVRequest* request)>;
+    void check_create_table(std::string instance_id, const CheckKVRequest* request,
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service.h:289]: It is not a public API, hide it in .cpp file

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 2]
```diff
@@ -614,4 +619,90 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+                                         std::string* msg, check_create_table_type get_check_info) {
+    std::unique_ptr<Transaction> txn;
+    TxnErrorCode err = txn_kv_->create_txn(&txn);
+    if (err != TxnErrorCode::TXN_OK) {
+        *code = cast_as<ErrCategory::READ>(err);
+        *msg = "failed to create txn";
+        return;
+    }
+    auto& [keys, hint, key_func] = get_check_info(request);
+
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:633]:
```suggestion
```

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 3]
```diff
@@ -614,4 +619,90 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+        return;
+    }
+    RPC_RATE_LIMIT(check_kv);
+    switch (request->op()) {
+    case CheckKVRequest::CREATE_INDEX_AFTER_FE_COMMIT: {
+        check_create_table(
+                instance_id, request, response, &code, &msg, [](const CheckKVRequest* request) {
+                    return std::make_tuple(request->check_keys().index_ids(), "index",
+                                           [](std::string instance_id, int64_t id) {
+                                               return recycle_index_key({instance_id, id});
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:688]:
```suggestion
                                               return recycle_index_key({std::move(instance_id), id});
```

**Comment by deardeng** [comment]
run buildall

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 4]
```diff
@@ -614,4 +621,93 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+            continue;
+        } else if (err == TxnErrorCode::TXN_OK) {
+            // find not match, prepare commit
+            *code = MetaServiceCode::ALREADY_EXISTED;
+            *msg = "prepare and commit rpc not match, recycle key remained";
+            return;
+        } else {
+            // err != TXN_OK, fdb read err
+            *code = cast_as<ErrCategory::READ>(err);
+            *msg = "ms read key error";
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:655]:
```suggestion
            *msg = fmt::format("ms read key error: {}", err);
```

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 5]
```diff
@@ -614,4 +621,93 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+            return;
+        }
+    }
+    LOG_INFO("check {} success request={}", hint, request->ShortDebugString());
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:659]: If there are many requests for a check, the request will be huge and unsuitable for printing directly to the log.

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 6]
```diff
@@ -614,4 +621,93 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+    return;
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:660]:
```suggestion
```

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 7]
```diff
@@ -614,4 +621,93 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+                    return std::make_tuple(
+                            request->check_keys().partition_ids(), "partition",
+                            [](std::string instance_id, int64_t id) {
+                                return recycle_partition_key({std::move(instance_id), id});
+                            });
+                });
+        break;
+    }
+    default:
+        DCHECK(false);
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service_partition.cpp:709]: It is more appropriate to return an invalid argument directly here.

**Comment by deardeng** [comment]
run buildall

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 8]
```diff
@@ -614,4 +621,94 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+    auto& [keys, hint, key_func] = get_check_info(request);
+    if (keys.empty()) {
+        *code = MetaServiceCode::INVALID_ARGUMENT;
+        *msg = "empty keys";
+        return;
+    }
+
+    for (auto id : keys) {
+        auto key = key_func(instance_id, id);
+        err = check_recycle_key_exist(txn.get(), key);
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_partition.cpp:644]: is it possible that txn may exceed 5s duration?
we can separate it to several txn for rubustness.

File: cloud/src/meta-service/meta_service_partition.cpp [hunk 9]
```diff
@@ -614,4 +621,94 @@ void MetaServiceImpl::drop_partition(::google::protobuf::RpcController* controll
...
+        if (err == TxnErrorCode::TXN_KEY_NOT_FOUND) {
+            continue;
+        } else if (err == TxnErrorCode::TXN_OK) {
+            // find not match, prepare commit
+            *code = MetaServiceCode::ALREADY_EXISTED;
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_partition.cpp:649]: this should be an internal error?

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 10]
```diff
@@ -549,6 +562,80 @@ public void commitMaterializedIndex(long dbId, long tableId, List<Long> indexIds
...
+        checkKvRequestBuilder.setCheckKeys(checkKeyInfosBuilder.build());
+        checkKvRequestBuilder.setOp(Cloud.CheckKVRequest.Operation.CREATE_PARTITION_AFTER_FE_COMMIT);
+        final Cloud.CheckKVRequest checkKVRequest = checkKvRequestBuilder.build();
+
+        Cloud.CheckKVResponse response = null;
+        int tryTimes = 0;
+        while (tryTimes++ < Config.metaServiceRpcRetryTimes()) {
+            try {
+                response = MetaServiceProxy.getInstance().checkKv(checkKVRequest);
+                if (response.getStatus().getCode() != Cloud.MetaServiceCode.KV_TXN_CONFLICT) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:584]: is `txn_confilict` already handled on the ms-server side? @w41ter ?

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 11]
```diff
@@ -549,6 +562,80 @@ public void commitMaterializedIndex(long dbId, long tableId, List<Long> indexIds
...
+                    break;
+                }
+            } catch (RpcException e) {
+                LOG.warn("tryTimes:{}, checkPartition RpcException", tryTimes, e);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:588]: what should we do if something wrong like this happened? 
it seems that the created table should be dropped.

File: fe/fe-core/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java [hunk 12]
```diff
@@ -627,6 +652,109 @@ public void executeDynamicPartition(Collection<Pair<Long, Long>> dynamicPartitio
...
+
+    private void cloudBatchBeforeCreatePartitions(boolean executeFirstTime,
+                                                  ArrayList<AddPartitionClause> addPartitionClauses,
+                                                  OlapTable olapTable, List<Long> indexIds, Database db,
+                                                  String tableName, List<Long> generatedPartitionIds)
+            throws DdlException {
+        if (Config.isNotCloudMode()) {
+            return;
+        }
+        if (executeFirstTime && !addPartitionClauses.isEmpty()) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java:728]:
```suggestion
        if (!executeFirstTime || addPartitionClauses.isEmpty()) return;
```

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 13]
```diff
@@ -1773,25 +1786,34 @@ public void addPartition(Database db, String tableName, AddPartitionClause addPa
...
+                    afterCreatePartitions(db.getId(), olapTable.getId(), partitionIds, indexIds, isCreateTable);
+                }
+                if (writeEditLog) {
+                    Env.getCurrentEnv().getEditLog().logAddPartition(info);
+                    LOG.info("succeed in creating partition[{}], temp: {}", partitionId, isTempPartition);
+                } else {
+                    LOG.info("postpone creating partition[{}], temp: {}", partitionId, isTempPartition);
+                }
+                if (!isCreateTable) {
+                    checkCreatePartitions(db.getId(), olapTable.getId(), partitionIds, indexIds);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:1814]: We have to check before writing edit log.

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 14]
```diff
@@ -2143,15 +2165,21 @@ protected Partition createPartitionWithIndices(long dbId, OlapTable tbl, long pa
...
             throws DdlException {
     }
 
-    protected void afterCreatePartitions(long dbId, long tableId, List<Long> partitionIds, List<Long> indexIds,
+    public void afterCreatePartitions(long dbId, long tableId, List<Long> partitionIds, List<Long> indexIds,
                                          boolean isCreateTable)
             throws DdlException {
     }
 
+    public void checkCreatePartitions(long dbId, long tableId, List<Long> partitionIds, List<Long> indexIds)
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:2178]: This function seems not necessary, the implementation can be done in `CloudInternalCatalog.afterCreatePartitions()`

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 15]
```diff
@@ -415,6 +416,18 @@ protected void afterCreatePartitions(long dbId, long tableId, List<Long> partiti
         }
     }
 
+    public void checkCreatePartitions(long dbId, long tableId, List<Long> partitionIds, List<Long> indexIds)
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:419]: This function seems not necessary, the implementation can be done in `CloudInternalCatalog.afterCreatePartitions()`

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: cloud/src/meta-service/meta_service_partition.cpp (hunk 9 above)
Comment (deardeng) [reply cloud/src/meta-service/meta_service_partition.cpp:649]: ALREADY_EXISTED

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java (hunk 11 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:588]: createtable catch exception，will drop failed table

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java (hunk 10 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:584]: MS encountered  txn_confilict will retry, and FE performed redundancy checking again. No significant impact

**Review by w41ter — DISMISSED** [review]

**Review by gavinchou — DISMISSED** [review]

**Comment by gavinchou** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by Yukang-Lian — APPROVED** [review]

**Review by ByteYue — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**